cmd/log/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+use humility::cli::Subcommand;
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Validate};
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list all stringbufs
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short, conflicts_with = "list")]
+    monitor: bool,
+    /// how long to delay in ms between polling for new log entries
+    #[clap(
+        long,
+        short,
+        default_value = "100",
+        requires = "monitor",
+        conflicts_with = "list"
+    )]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+fn stringbuf_read(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    last_last: u64,
+) -> Result<(String, u64)> {
+    let mut buf: Vec<u8> = vec![];
+    let mut log_msg: String = "".to_owned();
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    // Load the stringbuf data from buf
+    core.op_start()?;
+    core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+    core.op_done()?;
+
+    // use the buf to create a stringbuf struct
+    let ringbuf_val: Value = Value::Struct(
+        reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+    );
+    let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+    let stringbuf = Stringbuf::from_value(&cell.cell.value).unwrap();
+
+    // extract the log itself as a [u8]
+    let buffer = stringbuf.buffer.as_slice();
+    // the last written element in the log
+    let last: u64 = stringbuf.last.unwrap() as u64;
+
+    // read everything between current last, and last last by
+    // generating index range which connects these two in a circular manner, only including the
+    // new characters, and in the order they should be printed
+    // last_last has been printed already, last has not
+    let new_idxs: Vec<u64> = if last > last_last {
+        (last_last + 1..last).collect()
+    } else {
+        // This logic also works for last == last_last since we just need to print the entire range
+
+        // if last_last was not the last element of the buffer, then we include everything at
+        // the end of the ring buffer
+        if last_last as usize != buffer.len() - 1 {
+            // should not overflow the length of buffer as we just checked
+            let mut range = ((last_last + 1) as u64..buffer.len() as u64)
+                .collect::<Vec<u64>>();
+            range.append(&mut (0..=last).collect::<Vec<u64>>());
+            range
+        // last_last was the last element so we only need the beginning of the buffer
```

Comment:
This comment should be placed in the else block.

Reply:
moved.